File: src/capture/screencast.js
import fs from "node:fs";
import path from "node:path";
import * as ff from "../lib/ffmpeg.js";
import { log } from "../lib/log.js";

/**
 * CDP 화면 캐스트로 실제 디바이스 픽셀(예: 1080x1920) 프레임을 그대로 받아옵니다.
 * Playwright 내장 녹화는 CSS 픽셀 크기로만 찍혀서 세로 영상으로 확대하면 뿌옇게 나옵니다.
 */
export class Screencast {
  constructor(page, dir) {
    this.page = page;
    this.dir = path.join(dir, "frames");
    this.frames = [];
    this.client = null;
  }

  async start({ maxWidth, maxHeight, quality = 92 } = {}) {
    fs.mkdirSync(this.dir, { recursive: true });
    this.client = await this.page.context().newCDPSession(this.page);
    this.client.on("Page.screencastFrame", async (frame) => {
      const file = path.join(this.dir, `f_${String(this.frames.length).padStart(6, "0")}.jpg`);
      fs.writeFileSync(file, Buffer.from(frame.data, "base64"));
      this.frames.push({ file, ts: frame.metadata.timestamp ?? Date.now() / 1000, wall: Date.now() / 1000 });
      try { await this.client.send("Page.screencastFrameAck", { sessionId: frame.sessionId }); }
      catch { /* 페이지가 이미 닫혔으면 무시 */ }
    });
    await this.client.send("Page.startScreencast", {
      format: "jpeg", quality, maxWidth, maxHeight, everyNthFrame: 1,
    });
  }

  async stop() {
    try { await this.client?.send("Page.stopScreencast"); } catch { /* 이미 정리됨 */ }
  }

  /**
   * 프레임마다 실제 간격을 살려서 mp4 로 묶습니다.
   * 화면이 안 바뀌면 프레임이 안 오기 때문에 영상 길이는 실제 조작 시간보다 짧습니다.
   * 자막을 제자리에 붙이려면 "실제 시각 → 영상 시각" 대응표가 필요해서 같이 만들어 둡니다.
   */
  async assemble(outFile, { width, height, fps = 30, t0 = 0 }) {
    if (this.frames.length < 5) return null;
    const listFile = path.join(this.dir, "frames.txt");
    const lines = ["ffconcat version 1.0"];
    this.timeMap = [];
    let cursor = 0;
    for (let i = 0; i < this.frames.length; i++) {
      const next = this.frames[i + 1];
      const d = next ? Math.min(2, Math.max(1 / 60, next.ts - this.frames[i].ts)) : 1 / fps;
      this.timeMap.push({ wall: this.frames[i].wall - t0, video: cursor });
      cursor += d;
      lines.push(`file '${path.basename(this.frames[i].file)}'`, `duration ${d.toFixed(4)}`);
    }
    this.videoDuration = cursor;
    lines.push(`file '${path.basename(this.frames.at(-1).file)}'`);
    fs.writeFileSync(listFile, lines.join("\n"));

    await ff.run([
      "-f", "concat", "-safe", "0", "-i", listFile,
      "-vf", `fps=${fps},scale=${width}:${height}:force_original_aspect_ratio=decrease,pad=${width}:${height}:(ow-iw)/2:(oh-ih)/2:color=black`,
      "-c:v", "libx264", "-preset", "veryfast", "-crf", "18", "-pix_fmt", "yuv420p",
      outFile,
    ]);
    log.info(`화면 캐스트 ${this.frames.length}프레임 → ${path.basename(outFile)} (${cursor.toFixed(1)}초)`);
    return outFile;
  }

  /** 녹화 시작 후 n초(실제 시각)가 영상에서는 몇 초인지 환산합니다. */
  mapTime(wallSeconds) {
    const map = this.timeMap;
    if (!map?.length) return wallSeconds;
    if (wallSeconds <= map[0].wall) return 0;
    if (wallSeconds >= map.at(-1).wall) return this.videoDuration;
    let lo = 0, hi = map.length - 1;
    while (hi - lo > 1) {
      const mid = (lo + hi) >> 1;
      if (map[mid].wall <= wallSeconds) lo = mid; else hi = mid;
    }
    const span = map[hi].wall - map[lo].wall;
    const ratio = span > 0 ? (wallSeconds - map[lo].wall) / span : 0;
    return map[lo].video + (map[hi].video - map[lo].video) * ratio;
  }
}

File: python/sns_autopilot/capture/screencast.py
"""CDP 화면 캐스트로 실제 디바이스 픽셀 프레임을 받아 mp4 로 묶습니다.

Playwright 내장 녹화는 CSS 픽셀 크기로만 저장돼서 세로 영상으로 확대하면 뿌옇게 나옵니다.
대신 CDP `Page.startScreencast` 로 1080x1920 프레임을 직접 받습니다.
"""
from __future__ import annotations

import base64
import bisect
import time
from dataclasses import dataclass, field
from pathlib import Path

from .. import ffmpeg, log


@dataclass
class Frame:
    file: Path
    ts: float   # 프레임이 실제로 그려진 시각(epoch 초). 핸들러 실행 시각과 무관해 믿을 수 있습니다.


@dataclass
class Screencast:
    page: object
    dir: Path
    frames: list[Frame] = field(default_factory=list)
    client: object | None = None
    video_duration: float = 0.0
    _wall_axis: list[float] = field(default_factory=list)
    _video_axis: list[float] = field(default_factory=list)

    def __post_init__(self) -> None:
        self.dir = Path(self.dir) / "frames"

    def start(self, max_width: int, max_height: int, quality: int = 92) -> None:
        self.dir.mkdir(parents=True, exist_ok=True)
        self.client = self.page.context.new_cdp_session(self.page)
        self.client.on("Page.screencastFrame", self._on_frame)
        self.client.send(
            "Page.startScreencast",
            {"format": "jpeg", "quality": quality, "maxWidth": max_width,
             "maxHeight": max_height, "everyNthFrame": 1},
        )

    def _on_frame(self, frame: dict) -> None:
        file = self.dir / f"f_{len(self.frames):06d}.jpg"
        file.write_bytes(base64.b64decode(frame["data"]))
        metadata = frame.get("metadata") or {}
        self.frames.append(Frame(file=file, ts=metadata.get("timestamp") or time.time()))
        try:
            self.client.send("Page.screencastFrameAck", {"sessionId": frame["sessionId"]})
        except Exception:  # noqa: BLE001 - 페이지가 이미 닫혔으면 무시
            pass

    def stop(self) -> None:
        try:
            if self.client:
                self.client.send("Page.stopScreencast")
        except Exception:  # noqa: BLE001 - 이미 정리됨
            pass

    def assemble(self, out_file: Path, width: int, height: int, fps: int = 30, t0: float = 0.0) -> Path | None:
        """프레임마다 실제 간격을 살려 mp4 로 묶고, 자막용 시간 대응표를 만듭니다.

        화면이 안 바뀌면 프레임이 오지 않으므로 영상 길이는 실제 조작 시간보다 짧습니다.
        그래서 "실제 시각 → 영상 시각" 축을 함께 기록해 둡니다.
        """
        if len(self.frames) < 5:
            return None

        list_file = self.dir / "frames.txt"
        lines = ["ffconcat version 1.0"]
        cursor = 0.0
        self._wall_axis.clear()
        self._video_axis.clear()
        for index, frame in enumerate(self.frames):
            nxt = self.frames[index + 1] if index + 1 < len(self.frames) else None
            gap = min(2.0, max(1 / 60, nxt.ts - frame.ts)) if nxt else 1 / fps
            self._wall_axis.append(frame.ts - t0)
            self._video_axis.append(cursor)
            cursor += gap
            lines.append(f"file '{frame.file.name}'")
            lines.append(f"duration {gap:.4f}")
        lines.append(f"file '{self.frames[-1].file.name}'")
        list_file.write_text("\n".join(lines), encoding="utf-8")
        self.video_duration = cursor

        # fps 필터는 여기서 쓰지 않습니다. 화면이 멈춰 있던 구간(긴 duration)을 만나면
        # 앞부분을 통째로 버리고 start 를 밀어버려 영상 뒷부분이 잘립니다.
        # 프레임 간격을 그대로 살린 VFR 로 두고, 30fps 변환은 렌더 단계에서 합니다.
        ffmpeg.run([
            "-f", "concat", "-safe", "0", "-i", str(list_file),
            "-vf", (f"scale={width}:{height}:force_original_aspect_ratio=decrease,"
                    f"pad={width}:{height}:(ow-iw)/2:(oh-ih)/2:color=black"),
            "-fps_mode", "vfr",
            "-c:v", "libx264", "-preset", "veryfast", "-crf", "18", "-pix_fmt", "yuv420p",
            str(out_file),
        ])
        log.info(f"화면 캐스트 {len(self.frames)}프레임 → {out_file.name} ({cursor:.1f}초)")
        return out_file

    def map_time(self, wall_seconds: float) -> float:
        """녹화 시작 후 n초(실제 시각)가 영상에서는 몇 초인지 환산합니다."""
        axis = self._wall_axis
        if not axis:
            return wall_seconds
        if wall_seconds <= axis[0]:
            return 0.0
        if wall_seconds >= axis[-1]:
            return self.video_duration
        hi = bisect.bisect_right(axis, wall_seconds)
        lo = hi - 1
        span = axis[hi] - axis[lo]
        ratio = (wall_seconds - axis[lo]) / span if span > 0 else 0.0
        return self._video_axis[lo] + (self._video_axis[hi] - self._video_axis[lo]) * ratio
